File: Codechef/February challenge 2016/MTMXSUM.cpp
#include <iostream>
#include <stdio.h>

#define MOD 1000000007

using namespace std;

int main(){
	long long int N,K,i,j;
	scanf("%lld",&N);
	long long int A[100001],B[100001],temp1,temp2;
	temp1 = temp2 = 0;
	for(i = 0;i< N;i++){
		scanf("%lld",&A[i]);
		A[i] += i + 1;
		temp1 = ((A[i] % MOD) + (temp1%MOD))%MOD;
	}
	for(i = 0;i< N;i++){
		scanf("%lld",&B[i]);
		B[i] += i + 1;
		temp2 = ((B[i]%MOD) + (temp2%MOD))%MOD;
	}
	K = N;
	while(K > 0){
		printf("%lld ", ((temp2%MOD) * (temp1%MOD) % MOD);
		temp2 = temp1 = 0;
		for(j = 0;j < K - 1;j++){
			A[j] = max(A[j],A[j + 1]);
			//temp1 = ((A[j] % MOD) + (temp1%MOD))%MOD;
			temp1 += A[j] ;
			B[j] = max(B[j],B[j + 1]);
			// temp2 = ((B[j]%MOD) + (temp2%MOD))%MOD; 
			temp2 += B[j];
		}
		K--;
	}
	return 0;
}

File: Codechef/February challenge 2016/MTMXSUM.py
N = input()

var = 1000000007

A = map(int, raw_input().split())
B = map(int, raw_input().split())

list1 = []
list2 = []

for i in range(N):
	A[i] = A[i] + i + 1
	B[i] = B[i] + i + 1

print ((sum(A) % var) * (sum(B) % var) % var),

for i in range(N - 1):
	list1.append(max(A[i],A[i + 1]))
	list2.append(max(B[i],B[i + 1]))

K = N - 1

while(K > 0):
	# print list1
	# print list2
	# print K
	temp1 = sum(list1[:K]) % var
	temp2 = sum(list2[:K]) % var
	sum1 = (temp1 * temp2) % var
	print sum1,
	for i in range(K - 1):
		list1[i] = max(list1[i],list1[i + 1])
		list2[i] = max(list2[i],list2[i + 1])
	K -= 1